libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,159 @@
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADEN)) { }
+
+  // Once the ADC has been reset, enable it with the given settings
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    .ADC_Resolution = ADC_Resolution_12b,
+    .ADC_ContinuousConvMode = adc_mode,
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Set the ADC to wait mode
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, ENABLE);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.current_channel = 0;
+
+  // Start background conversions if in continuous
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= !(1 << adc_channel);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(ADCChannel adc_channel, ADCCallback callback, void *context) {
+  // Returns invalid if the given address is not connected to an ADC channel
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
+  s_adc_interrupts[adc_channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+uint16_t adc_read_value(ADCChannel adc_channel) {
+  if (!s_adc_status.continuous) {
+    s_adc_status.current_channel = 0;
+    ADC_StartOfConversion(ADC1);
+    while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) {}
+  }
+
+  return s_adc_interrupts[adc_channel].reading;
+}
+
+void ADC1_COMP_IRQHandler() {
+  ADCChannel current_channel = s_adc_status.current_channel;
+  uint32_t select = ADC1->CHSELR;
+  uint16_t reading;
+
+  if (select & (1 << current_channel)) {
+    reading = ADC_GetConversionValue(ADC1);
+    if (s_adc_interrupts[current_channel].callback != NULL) {
+      s_adc_interrupts[current_channel].callback(current_channel, reading,
+                                          s_adc_interrupts[current_channel].context);
+    }
+    s_adc_interrupts[current_channel].reading = reading;
+  }
+
+  current_channel = prv_get_ffs(select, current_channel);
```

Comment:
By definition, `ffs`, given an integer, returns the index of the first set bit in the integer. It doesn't seem like this is what `prv_get_ffs` is doing. (I still recommend using the intrinsic instead.)